Tools/CISupport/Shared/de-duper

```diff
@@ -0,0 +1,142 @@
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import argparse
+import os
+import subprocess
+import shutil
+import sys
+
+ROOT = '/var/results'
+STORAGE = 'storage'
+
+
+def parse_args(args):
+    parser = argparse.ArgumentParser(
+        description='Program which de-duplicates PNG files in a directory using hard links'
+    )
+    parser.add_argument(
+        'root', nargs='?',
+        help='Path to root of directory to de-duplicate PNG files in',
+        type=str, default='.',
+    )
+    parser.add_argument(
+        '-s', '--storage',
+        help='Directory to store images the program is linking to',
+        type=str,
+        default='storage',
+    )
+    return parser.parse_args(args)
+
+
+def find(path, filter):
+    process = None
+
+    try:
+        process = subprocess.Popen([
+                'find', path,
+                '-name', filter,
+            ], stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            encoding='utf-8',
+        )
+
+        line = process.stdout.readline()
+        while line:
+            yield line.rstrip()
+            line = process.stdout.readline()
+
+    finally:
+        if process and process.poll() is None:
+            process.kill()
+
+
+def md5(path):
+    result = subprocess.run(
+        ['md5sum', path],
+        encoding='utf-8',
+        capture_output=True
+    )
+    if result.returncode:
+        return None
+    return result.stdout.split()[0]
+
+
+def is_linked(path):
+    return os.stat(path).st_nlink > 1
+
+
+def main(args):
+    args = parse_args(args)
+
+    args.root = os.path.abspath(args.root)
+    args.storage = os.path.abspath(args.storage) if args.storage[0] in ('.', '/') else os.path.join(args.root, args.storage)
+    if not os.path.isdir(args.root):
+        sys.stderr.write(f"'{args.root}' is a path which does not exist\n")
+        return 1
+    storage_parent = os.path.dirname(args.storage)
+    if not os.path.isdir(storage_parent):
+        sys.stderr.write(f"'{storage_parent}' is a path which does not exist\n")
+        return 1
+
+    if not os.path.isdir(args.storage):
+        os.mkdir(args.storage)
+
+    cleaned_up = 0
+    de_duped = 0
+    processed = 0
+
+    existing_links = set()
+    for existing_file in os.listdir(args.storage):
```

Comment:
We should be using `os.scandir`, then we don't need to have extra syscalls and potentially reopen the directory from disk to call `os.stat`.